### ISSUE TITLE
Question: Clause Ordering In TRACECHECK Resolution Chains

### ISSUE THREAD

**Issue by alex-ozdemir** [benjaminkiesl/drat2er#3]
When `drat2er` produces TRACECHECK output, it prints a sequence of resolutions chains.

In my experience so far, the clauses in any resolution chain produce by `drat2er` have always been in resolution-order (i.e. if the chain is 1 3 5 4, then 1 and 3 a resolved first, and their result is resolved with 5 and that result is in turn resolved with 4).

Is this always the case?

**Comment by benjaminkiesl** [comment]
Hi Alex,

The way things are currently implemented, this should indeed always be the case (for more information, see below). But there is no software contract or any explicit unit tests meant to guarantee that it is indeed the case (since, at least until now, I never considered this a necessity).

Here is some more info on why it is indeed the case:

1. Resolvents are generated in the function Formula::DeriveSubsumingClause(const Clause& rup) where the indices of the resolution partners (called the "positive hints") are added to a vector in the order in which the resolutions are performed. 

2. When such a resolvent is written to the output by the function ToLRAT(const RupClause& rup)  (which is called by the function ToTRACECHECK(const RupClause& rup) in the file instruction_serialization.cc), the positive hints are printed in the order in which they occur in the vector.

So basically things could, for instance, change if someone would change the data structure for the positive hints to some unordered data structure in the future. But at this point I see no reason why anyone would want to do this.

Does this answer your question? Or do you maybe have good reasons why this property is useful and why it should maybe be guaranteed by drat2er?

Best wishes,
Benjamin

**Comment by alex-ozdemir** [comment]
Whoops!

I thought I had commented.

Anyways, Hi Ben! Thanks for responding so rapidly!

Your explanation makes sense. The use-case for resolution-ordered clauses in resolution chains is somewhat narrow: it makes it easier to construct an actual resolution tree.

Cheers!
Alex
